flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/partition/JobAwareShuffleEnvironment.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor.partition;
+
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.metrics.MetricGroup;
+import org.apache.flink.runtime.deployment.InputGateDeploymentDescriptor;
+import org.apache.flink.runtime.deployment.ResultPartitionDeploymentDescriptor;
+import org.apache.flink.runtime.executiongraph.ExecutionAttemptID;
+import org.apache.flink.runtime.executiongraph.PartitionInfo;
+import org.apache.flink.runtime.io.network.api.writer.ResultPartitionWriter;
+import org.apache.flink.runtime.io.network.partition.PartitionProducerStateProvider;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.shuffle.ShuffleEnvironment;
+import org.apache.flink.runtime.taskexecutor.TaskExecutor;
+
+import java.io.IOException;
+import java.util.Collection;
+
+/**
+ * Job-aware variation of the {@link ShuffleEnvironment} interface.
+ *
+ * @see ShuffleEnvironment
+ */
+public interface JobAwareShuffleEnvironment<P extends ResultPartitionWriter, G extends InputGate> extends AutoCloseable {
+
+	/**
+	 * Start the internal related services upon {@link TaskExecutor}'s startup.
+	 *
+	 * @return a port to connect to the task executor for shuffle data exchange, -1 if only local connection is possible.
+	 */
+	int start() throws IOException;
```

Comment:
The methods which already exist in `ShuffleEnvironment` do not need to be defined here, and we could make `JobAwareShuffleEnvironmentImpl` implements `JobAwareShuffleEnvironment`, `ShuffleEnvironment` to solve this issue.

Reply:
Unfortunately no, then we'd introduce the risk that someone uses the wrong method, like createPartitionWriters without the JobID.

Reply:
That said, if we were to introduce a JobID parameter into `ShuffleEnvironment#createResultPartitionWriters` we could probably just extend it. Long-term we need `releasePartition` methods without a JobID parameter anyway for global partitions.

Reply:
Hmm, my previous thought was putting `JobID` into shuffle context which was proposed by Andrey in another PR via `ShuffleEnvironment#createResultPartitionWriters(context)`, then in `JobAwareShuffleEnvironmentImpl` it could get job id from context to avoid two `createResultPartitionWriters` methods separately.

Reply:
yes, we could also pass the JobID into the context.

For the time-being I'll update the PR so that `JobAwareShuffleEnvironmentImpl extends ShuffleEnvironment`, add add a `createResultPartitionWriters` method that accepts a JobID.

Once the context is merged we can migrate to that.